internal/engine/local/servercontroller.go

```diff
@@ -0,0 +1,197 @@
+	defer st.RUnlockState()
+
+	// Find all the Cmds owned by CmdServer.
+	//
+	// Simulates controller-runtime's notion of Owns().
+	ownedCmds := make(map[string]*Cmd)
+	for _, cmd := range state.Cmds {
+		for _, ref := range cmd.ObjectMeta.OwnerReferences {
+			if ref.Kind != "CmdServer" {
+				continue
+			}
+
+			ownedCmds[ref.Name] = cmd
+		}
+	}
+
+	var r []CmdServer
+
+	// Infer all the CmdServer objects from the legacy EngineState
+	for _, mt := range state.Targets() {
+		if !mt.Manifest.IsLocal() {
+			continue
+		}
+		lt := mt.Manifest.LocalTarget()
+		if lt.ServeCmd.Empty() ||
+			mt.State.LastSuccessfulDeployTime.IsZero() {
+			continue
+		}
+
+		name := mt.Manifest.Name.String()
+		cmdServer := CmdServer{
+			ObjectMeta: ObjectMeta{
+				Name: name,
+			},
+			Spec: CmdServerSpec{
+				Args:           lt.ServeCmd.Argv,
+				Dir:            lt.ServeCmd.Dir,
+				Env:            lt.ServeCmd.Env,
+				TriggerTime:    mt.State.LastSuccessfulDeployTime,
+				ReadinessProbe: lt.ReadinessProbe,
+			},
+		}
+
+		cmd, ok := ownedCmds[mt.Manifest.Name.String()]
+		if ok {
+			cmdServer.Status = CmdServerStatus{
+				CmdName:   cmd.Name,
+				CmdStatus: cmd.Status,
+			}
+		}
+
+		r = append(r, cmdServer)
+	}
+
+	return r
+}
+
+func (c *ServerController) reconcile(ctx context.Context, server CmdServer, st store.RStore) {
+	cmdSpec := CmdSpec{
+		Args:           server.Spec.Args,
+		Dir:            server.Spec.Dir,
+		Env:            server.Spec.Env,
+		ReadinessProbe: server.Spec.ReadinessProbe,
+	}
+
+	name := server.Name
+	created, isCreated := c.createdCmds[name]
+	triggerTime := c.createdTriggerTime[name]
+	if isCreated && equality.Semantic.DeepEqual(created.Spec, cmdSpec) && triggerTime.Equal(server.Spec.TriggerTime) {
+		// We're in the correct state! Nothing to do.
+		return
+	}
+
+	// Otherwise, we need to create a new command.
+
+	// If the command hasn't appeared yet, wait until it appears.
+	if isCreated && server.Status.CmdName == "" {
+		return // wait for the command to appear
+	}
+
+	// If the command is running, wait until it's deleted
+	if isCreated && server.Status.CmdStatus.Terminated == nil {
+		if !c.deletingCmds[name] {
+			c.deletingCmds[name] = true
+			st.Dispatch(CmdDeleteAction{Name: server.Status.CmdName})
+		}
+		return
+	}
+
+	// Start the command!
+	c.createdTriggerTime[name] = server.Spec.TriggerTime
+	c.deletingCmds[name] = false
+	c.cmdCount++
+
+	cmdName := fmt.Sprintf("%s-serve-%d", name, c.cmdCount)
+	spanID := SpanIDForServeLog(c.cmdCount)
+
+	cmd := &Cmd{
+		ObjectMeta: ObjectMeta{
+			Name: cmdName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name: name,
+					Kind: "CmdServer",
+				},
+			},
+			Annotations: map[string]string{
+				v1alpha1.AnnotationSpanID: string(spanID),
+			},
+			Labels: map[string]string{
+				v1alpha1.LabelManifest: name,
+			},
+		},
+		Spec: cmdSpec,
+	}
+	c.createdCmds[name] = cmd
+	st.Dispatch(CmdCreateAction{Cmd: cmd})
+}
+
+type CmdServer struct {
```

Comment:
Will these become API server types in the future?

Reply:
ya, i think so - but still figuring out what the Status field should look like